### PR TITLE
examples: Add GC tracing example

### PR DESCRIPTION
**Pull request by tynopex**
This shows how to store GC pointers within the embedding and how to
trace custom and existing types.

### CODE REVIEW

File: examples/tracing.cpp [hunk 1]
```diff
@@ -0,0 +1,153 @@
+#include <memory>
+
+#include <jsapi.h>
+
+#include "boilerplate.h"
+
+// This example illustrates how to safely store GC pointers in the embedding's
```
Comment (hotsphink) [inline examples/tracing.cpp:8]: I think this should probably specify that this applies to regular strong references only (without getting into too much detail.) Weak references require different goop.

File: examples/tracing.cpp [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+// data structures by implementing appropriate tracing mechanisms.
+
+////////////////////////////////////////////////////////////
+
+// An example C++ type that stores an arbitrary JS value.
+struct SafeBox {
+  // Arbitrarily JS value that will need to be traced. The JS::Heap type has a
```
Comment (hotsphink) [inline examples/tracing.cpp]: *Arbitrary

File: examples/tracing.cpp [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+  // constructor and destructor and will trigger any required GC barriers on
```
Comment (hotsphink) [inline examples/tracing.cpp]: It does indeed have a constructor and destructor -- maybe say that they need to run in order to properly register the pointer with the GC?

File: examples/tracing.cpp [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+  // If we use SafeBox as a stack object, then a JS::Rooted is enough.
+  JS::Rooted<SafeBox> stackSafe(cx);
+
+  // We can also use js::UniquePtr if SafeBox should be allocated on heap.
+  JS::Rooted<js::UniquePtr<SafeBox>> heapSafe(cx, js::MakeUnique<SafeBox>());
+
+  // NOTE: A JS::Rooted<SafeBox*> is a compile error. If one wanted to support
+  // rooting bare non-GC pointers then both JS::MapTypeToRootKind and
+  // JS::GCPolicy need to be defined for SafeBox*. This should be avoided in
+  // favor of using a smart-pointer when possible.
```
Comment (hotsphink) [inline examples/tracing.cpp:48]: Ugh, that comment nerd-sniped me. Assuming it pleaseth the Masterful Waldo, the MapTypeToRootKind requirement should go away soon. (Which doesn't matter in this repo for now.)

File: examples/tracing.cpp [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+  return true;
+}
+
+////////////////////////////////////////////////////////////
+
+// By defining a JS::GCPolicy we can support tracing in existing types that we
```
Comment (hotsphink) [inline examples/tracing.cpp]: s/in //

File: examples/tracing.cpp [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+      GCPolicy<T>::trace(trc, target, name);
+    }
+  }
+  static bool needsSweep(std::shared_ptr<T>* tp) {
+    if (T* target = tp->get()) {
+      return GCPolicy<T>::needsSweep(target);
+    }
+    return false;
+  }
+  static bool isValid(const std::shared_ptr<T>& t) {
```
Comment (hotsphink) [inline examples/tracing.cpp:73]: ...and it looks like our docs don't include this isValid() requirement. Bug 1534440.

File: examples/tracing.cpp [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+  // We can use std::shared_ptr too since we defined a GCPolicy above.
+  JS::Rooted<std::shared_ptr<SafeBox>> sharedSafe(cx,
+                                                  std::make_shared<SafeBox>());
+  return true;
+}
+
+////////////////////////////////////////////////////////////
+
+// When an embedding wishes to keep GC things alive even if the JavaScript no
+// longer cares has direct references, it must provide GC roots for the various
```
Comment (hotsphink) [inline examples/tracing.cpp]: s/cares //.

s/even if/even though/? Or s/even if/when/?

File: examples/tracing.cpp [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+// tracing mechanisms to search from. This is done using the PersistentRooted
+// type.
+//
+// Using a custom trace function can allow you to use a single PersistentRooted
+// to traverse the embeddings private state holding many GC pointers.
```
Comment (hotsphink) [inline examples/tracing.cpp]: Why is it that are you doing the thing of being so passive? "A custom trace functions allows a single PersistentRooted..." or "Use a custom trace function to enable a single..." or something.

Also, s/embeddings/embedding's/.

File: examples/tracing.cpp [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+
+// A global PersistentRooted is created before SpiderMonkey has initialized so
+// we must be careful that we do not create any JS::Heap fields during
```
Comment (hotsphink) [inline examples/tracing.cpp]: I'll let you get away with the passive voice here, but at least s/that we do not/to not/.

File: examples/tracing.cpp [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+static bool GlobalRootExample(JSContext* cx) {
+  // Initialize the root with cx and allocate a fresh SafeBox.
+  globalPtrSafe.init(cx, new SafeBox);
+
+  // If we want to avoid a heap allocation, we can wrap the PersistentRooted in
+  // a Maybe. When we emplace the variable, we pass 'cx' for the constructor of
+  // PersistentRooted.
+  globalMaybeSafe.emplace(cx);
+
+  // IMPORTANT: When using global PersistentRooted they *must* be cleared
```
Comment (hotsphink) [inline examples/tracing.cpp]: "PersistentRooteds", I guess, to make the plurality agree with "they"?

File: examples/tracing.cpp [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+  // before shutdown by calling 'reset()'.
+  globalMaybeSafe.reset();
+  globalPtrSafe.reset();
+
+  return true;
+}
+
+////////////////////////////////////////////////////////////
+
+// Instead of the global variable example above, it is often preferrable to
```
Comment (hotsphink) [inline examples/tracing.cpp]: *preferable

**Review by hotsphink — APPROVED** [review]
Great examples! Thanks!

**Comment by tynopex** [comment]
Thanks for the editorial feedback. I'll land this once #18 lands.

File: examples/tracing.cpp [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+  // Initialize the root with cx and allocate a fresh SafeBox.
+  globalPtrSafe.init(cx, new SafeBox);
+
+  // If we want to avoid a heap allocation, we can wrap the PersistentRooted in
+  // a Maybe. When we emplace the variable, we pass 'cx' for the constructor of
+  // PersistentRooted.
+  globalMaybeSafe.emplace(cx);
+
+  // IMPORTANT: When using global PersistentRooteds they *must* be cleared
+  // before shutdown by calling 'reset()'j
```
Comment (ptomato) [inline examples/tracing.cpp]: extraneous 'j'

File: meson.build [hunk 13]
```diff
@@ -67,3 +67,4 @@ add_project_arguments(cxx.get_supported_arguments(test_warning_args),
 executable('hello', 'examples/hello.cpp', 'examples/boilerplate.cpp', dependencies: spidermonkey)
 executable('cookbook', 'examples/cookbook.cpp', 'examples/boilerplate.cpp', dependencies: spidermonkey)
 executable('repl', 'examples/repl.cpp', 'examples/boilerplate.cpp', dependencies: [spidermonkey, readline])
+executable('tracing', 'examples/tracing.cpp', 'examples/boilerplate.cpp', dependencies: [spidermonkey, readline])
```
Comment (ptomato) [inline meson.build]: This doesn't need to link to readline, `dependencies: spidermonkey` should be fine.

File: examples/tracing.cpp [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+// A global PersistentRooted is created before SpiderMonkey has initialized so
+// we must be careful to not create any JS::Heap fields during construction.
+JS::PersistentRooted<js::UniquePtr<SafeBox>> globalPtrSafe;
+mozilla::Maybe<JS::PersistentRooted<SafeBox>> globalMaybeSafe;
+
+static bool GlobalRootExample(JSContext* cx) {
+  // Initialize the root with cx and allocate a fresh SafeBox.
+  globalPtrSafe.init(cx, new SafeBox);
+
+  // If we want to avoid a heap allocation, we can wrap the PersistentRooted in
```
Comment (ptomato) [inline examples/tracing.cpp:110]: Just curious, are there any other considerations to use one or the other of these, besides avoiding a heap allocation?

**Review by ptomato — APPROVED** [review]
I had a couple of minor comments but overall this looks really useful!

File: meson.build (hunk 13 above)
Comment (tynopex) [reply meson.build]: Good catch!

File: examples/tracing.cpp (hunk 14 above)
Comment (tynopex) [reply examples/tracing.cpp:110]: Using a Maybe is really only for avoiding a heap allocation.

**Comment by tynopex** [comment]
I updated example to have a std::vector<JS::Heap> as well to clarify the point that Root<Array> is preferred over Array<Root>.

**Pull request merged**